Further refinement of {Float,Double,Float80}.init(_:String)

tbkka opened the pull request:
After talking with @stephentyrone, I found some additional simplifications. No
functional change, just shorter/simpler.

tbkka committed:
Further refinement of {Float,Double,Float80}.init(_:String)

After talking with @stephentyrone, I found some additional simplifications. No
functional change, just shorter/simpler.

tbkka commented:
@swift-ci Please smoke test Linux macOS

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please smoke test macOS

airspeedswift commented on stdlib/public/core/FloatingPointParsing.swift.gyb:
```diff
@@ -126,36 +126,23 @@ extension ${Self}: LosslessStringConvertible {
+        default:
+          break
+        }
+        let endPtr = _swift_stdlib_strto${cFuncSuffix2[bits]}_clocale(chars, p)
```
`_swift_stdlib_strto...` has existed since 5.0, right?

airspeedswift replied:
oh ignore me, it was there already

stephentyrone replied:
Yes, no ABI change here.

airspeedswift commented on stdlib/public/core/FloatingPointParsing.swift.gyb:
```diff
@@ -126,36 +126,23 @@ extension ${Self}: LosslessStringConvertible {
   ///   is `nil`.
   @inlinable // FIXME(sil-serialize-all)
```
thinking of this comment (that we should probably remove...) could we use `@_specialize` and then make this non-inlinable to reduce having to inline it?

airspeedswift replied:
given there are only 2 string protocol conformance rn

stephentyrone replied:
Not unreasonable.

jckarter replied:
Even better, make the generic thing a stub that just canonicalizes to `Substring` and make the non-inlinable implement take `Substring`. (Which is almost always better in every respect than using `StringProtocol`)

tbkka replied:
Do you mean like this?
```
  @inlinable
  public init?(_ text: String) {
    self.init(substring: text[...])
  }

  public init?(substring: Substring) {
    ... real work ...
  }
```

jckarter replied:
Yeah.

tbkka replied:
Done.

tbkka committed:
Merge branch 'master' into tbkka-floatparse-simplify-2

tbkka committed:
Further reduce inlined code size.

This makes the generic inlineable part a small stub that delegates to the full
non-inlined version.

ABI compatibility:
* We support the same generic init() as before
* _swift_stdlib_strtoXYZ_clocale is still available to
  support old inlined code

API addition:
* We now have a public specialized form of init?(_: Substring)
  in addition to the generic init?<S:StringProtocol> form.

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please smoke test macOS

tbkka committed:
Add @available marker to new API

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please smoke test macOS

tbkka committed:
Support back-deployment to older OSes by inlining the full version

tbkka commented:
Whew!  I think I _finally_ worked out how to correctly handle the backwards- and forwards- deployment issues here.

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
The CI failed, but looking through the log I don't understand why.  I see no build or test failures.  I do see one inscrutable message:
```
10:18:37 E[10:18:37.417] Transport error: Input/output error
```

jckarter commented:
@tbkka You might just try again, I'd judge that to be a transient failure.

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
Now the buildbot is failing with a missing symbol `_$sSdySdSgSscfC` -- I'm not yet very good at demangling in my head 😉 but I think this is the new Double constructor I added.  Not sure why it would only fail on this one test, but it seems that I'm not using the right availability declarations?

```
18:42:46 Test Case '-[SourceKitTests.DocumentColorTests testSimple]' started.
18:42:46 dyld: lazy symbol binding failed: Symbol not found: _$sSdySdSgSscfC
18:42:46   Referenced from: /Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/buildbot_incremental/sourcekitlsp-macosx-x86_64/x86_64-apple-macosx/release/SourceKitLSPPackageTests.xctest/Contents/MacOS/SourceKitLSPPackageTests
18:42:46   Expected in: /usr/lib/swift/libswiftCore.dylib
```

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
Hmmm....  There's something wrong with how sourcekit-lsp is being tested here:  Some part of that test suite is being built against the new standard library (hence is referencing the new symbol) but is then getting run with the system-installed standard library (the reference to /usr/lib/swift above).

benlangmuir commented:
> Hmmm.... There's something wrong with how sourcekit-lsp is being tested here: Some part of that test suite is being built against the new standard library (hence is referencing the new symbol) but is then getting run with the system-installed standard library (the reference to /usr/lib/swift above).

This sounds expected to me. We cannot use libswiftCore.dylib from the toolchain on macOS, we have to use the system one at runtime.  Otherwise we couldn't use anything else from the system (e.g. Foundation) without creating a conflict.

futurejones commented:
Can one of the admins verify this patch?

tbkka commented:
@swift-ci Please smoke test macOS

tbkka commented:
I haven't been able to reproduce this particular test failure locally, so I'll need to keep poking swift-ci to run it for me.

tbkka commented:
@swift-ci Please smoke test macOS

shahmishal commented:
@swift-ci Please smoke test macOS

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
CI tests are all succeeding now, so I went ahead and merged.
